### DISCUSSION TITLE
Add Firefox as a source

### DISCUSSION THREAD

**Discussion by Dunexus** [ImranR98/Obtainium#1484] [Ideas]
Hello,

Since Firefox 124 (18 march), Mozilla is deprecating its github repository for android and will stop shipping github release after 126.

The apks can still available through their ftp : https://ftp.mozilla.org/pub/fenix/releases/

However, it is way more complex to configure as a source than a github repo. Maybe we could add a preconfigured Firefox source to easy the installation on it through Obtainium.

**Comment by ImranR98** [comment]
It's not a good idea to add something as a custom source if it already works using the HTML source. Doing so would lead to a lot of bloat over time and reduce the maintainability of the app.

Instead, one person could figure out how to add this to Obtainium and share their settings on https://apps.obtainium.imranr.dev/

**Reply by Dunexus** [reply]
I've been using Obtainium for over a year  and I just learn about this sharing website !
Thanks, I will try to find a way to configure firefox using html

**Reply by Dunexus** [reply]
Spent over an hour and didn't manage to get it working.
Html mode doesn't seem to handle hierarchy of links as the ftp provides.
Maybe someone else will find a way.

**Reply by laxmanpradhan** [reply]
> It's not a good idea to add something as a custom source if it already works using the HTML source. Doing so would lead to a lot of bloat over time and reduce the maintainability of the app.
> 
> Instead, one person could figure out how to add this to Obtainium and share their settings on https://apps.obtainium.imranr.dev/

@ImranR98 : Using the Firefox configuration from this site, it doesn't get the correct sub version.

Currently 127.0.2 is the latest however it all says 127.0 is the latest and gives an error saying that it can't update because it can't install the older version.I have 127.0.1 currently installed.

I think the regex is only looking for the first two numbers in the version and should be:

Link1: `[0-9]+\.[0-9]+\.[0-9]/$`

Link 3: `fenix-[0-9]+\.[0-9]+\.[0-9]+-android-arm64-v8a/$`

Version string extraction: `[0-9]+\.[0-9]+\.[0-9]`

But I'm not sure what will happen when there are only 2 number or 128.0 as the next release.

Perhaps something like for link 1 this which blacklists any letters and makes the last number optional, however obtainium gives and error saying no suitable release:

`(?!.*b)[0-9]+\.[0-9]+\.?[0-9]?`

In the above, adding the exclusion criteria `(?!.*b)` works as expected, letter b will not be matched (no beta versions). The optional match for the third subversion number is what doesn't work `\.?[0-9]?` but I'm not sure why as it seems like it should in regex101:

![image](https://github.com/ImranR98/Obtainium/assets/297760/c6cb7ccb-57ba-4966-a0e2-628a7b633604)


But Obtainium gives an error `no suitable release`:
![Screenshot_20240627-152932](https://github.com/ImranR98/Obtainium/assets/297760/171f49e8-3002-4438-8cf1-2e68e0770cef)

**Comment by DwainZwerg** [comment]
> Spent over an hour and didn't manage to get it working.

I've already spent hours on this because I use Firefox Nightly, which was not available on GitHub at any time (see also https://github.com/ImranR98/Obtainium/discussions/621). I've come to terms with the fact that I can only update this app like AuroraStore.

**Comment by rvandermeulen** [comment]
In case it helps, https://product-details.mozilla.org/1.0/mobile_versions.json is a supported endpoint for determining what the latest versions are for each channel, maybe used in combination with the previously-given FTP link. Should be useful for Release and Beta anyway. Nightly is a bit more challenging. There are places within the Mozilla CI system where you could always find the latest-available build, but I don't know how easily that'd integrate with this tool.

**Comment by rabaimorp** [comment]
Just adding a direct apk link to the latest arm64-v8a build from the oficial repo worked fine for me. Auto updates work as well, since the link to the latest apk is always the same.

https://apps.obtainium.imranr.dev/redirect?r=obtainium://app/%7B%22id%22%3A%22org.mozilla.fenix%22%2C%22url%22%3A%22https%3A%2F%2Ffirefox-ci-tc.services.mozilla.com%2Fapi%2Findex%2Fv1%2Ftask%2Fmobile.v3.firefox-android.apks.fenix-nightly.latest.arm64-v8a%2Fartifacts%2Fpublic%252Fbuild%252Ftarget.arm64-v8a.apk%22%2C%22author%22%3A%22firefox-ci-tc.services.mozilla.com%22%2C%22name%22%3A%22Firefox%20Nightly%22%2C%22preferredApkIndex%22%3A0%2C%22additionalSettings%22%3A%22%7B%5C%22requestHeader%5C%22%3A%5B%7B%5C%22requestHeader%5C%22%3A%5C%22User-Agent%3A%20Mozilla%2F5.0%20(Linux%3B%20Android%2010%3B%20K)%20AppleWebKit%2F537.36%20(KHTML%2C%20like%20Gecko)%20Chrome%2F114.0.0.0%20Mobile%20Safari%2F537.36%5C%22%7D%5D%2C%5C%22trackOnly%5C%22%3Afalse%2C%5C%22invertAPKFilter%5C%22%3Afalse%2C%5C%22appName%5C%22%3A%5C%22%5C%22%2C%5C%22shizukuPretendToBeGooglePlay%5C%22%3Afalse%2C%5C%22exemptFromBackgroundUpdates%5C%22%3Afalse%2C%5C%22skipUpdateNotifications%5C%22%3Afalse%2C%5C%22about%5C%22%3A%5C%22%5C%22%2C%5C%22versionDetection%5C%22%3Afalse%2C%5C%22autoApkFilterByArch%5C%22%3Afalse%7D%22%7D

**Comment by Dunexus** [comment]
Looks like someone managed to get it working and shared a conf on https://apps.obtainium.imranr.dev/

Not sure who they are but huge thanks

**Reply by DwainZwerg** [reply]
The point is that until recent times there was a missing feature in Obtainium allowing us to include the Mozilla repo (see https://github.com/ImranR98/Obtainium/issues/1625#issuecomment-2128191552). This was fixed by @ImranR98, so that our (on the whole correct) configs finally worked. Now, I (triggering several json errors, which is why the site was down for the last two days) and noodlejetski have added the configs to the website.

So you can close this discussion as resolved.
